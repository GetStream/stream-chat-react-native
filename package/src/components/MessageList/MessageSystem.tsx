import React from 'react';
import { StyleProp, StyleSheet, Text, View, ViewStyle } from 'react-native';

import type { MessageType } from './hooks/useMessageList';

import { useTheme } from '../../contexts/themeContext/ThemeContext';
import {
  isDayOrMoment,
  TDateTimeParserInput,
  useTranslationContext,
} from '../../contexts/translationContext/TranslationContext';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
    marginBottom: 10,
  },
  line: {
    flex: 1,
    height: 0.5,
  },
  text: {
    fontSize: 10,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  textContainer: {
    flex: 3,
    marginTop: 10,
  },
});

export type MessageSystemProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  /** Current [message object](https://getstream.io/chat/docs/#message_format) */
  message: MessageType<At, Ch, Co, Ev, Me, Re, Us>;
  /**
   * Formatter function for date object.
   *
   * @param date TDateTimeParserInput object of message
   * @returns string
   */
  formatDate?: (date: TDateTimeParserInput) => string;
  style?: StyleProp<ViewStyle>;
};

/**
 * A component to display system message. e.g, when someone updates the channel,
 * they can attach a message with that update. That message will be available
 * in message list as (type) system message.
 */
export const MessageSystem = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  props: MessageSystemProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { formatDate, message, style } = props;

  const {
    theme: {
      colors: { grey, grey_whisper },
      messageList: {
        messageSystem: { container, dateText, line, text, textContainer },
      },
    },
  } = useTheme();
  const { tDateTimeParser } = useTranslationContext();

  const createdAt = message.created_at;
  const parsedDate = tDateTimeParser(createdAt);
  const date =
    formatDate && createdAt
      ? formatDate(createdAt)
      : parsedDate && isDayOrMoment(parsedDate)
      ? parsedDate.calendar().toUpperCase()
      : parsedDate;

  return (
    <View style={[styles.container, style, container]} testID='message-system'>
      <View style={[styles.line, { backgroundColor: grey_whisper }, line]} />
      <View style={[styles.textContainer, textContainer]}>
        <Text style={[styles.text, { color: grey }, text]}>
          {message.text?.toUpperCase() || ''}
        </Text>
        <Text style={[styles.text, { color: grey }, dateText]}>{date}</Text>
      </View>
      <View style={[styles.line, { backgroundColor: grey_whisper }, line]} />
    </View>
  );
};

MessageSystem.displayName = 'MessageSystem{messageList{messageSystem}}';
