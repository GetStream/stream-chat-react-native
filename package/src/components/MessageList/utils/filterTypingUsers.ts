import type { ChatContextValue } from '../../../contexts/chatContext/ChatContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type { TypingContextValue } from '../../../contexts/typingContext/TypingContext';
import type { DefaultStreamChatGenerics } from '../../../types/types';

type FilterTypingUsersParams<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Pick<TypingContextValue<StreamChatGenerics>, 'typing'> &
  Pick<ChatContextValue<StreamChatGenerics>, 'client'> &
  Pick<ThreadContextValue<StreamChatGenerics>, 'thread'>;

export const filterTypingUsers = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  client,
  thread,
  typing,
}: FilterTypingUsersParams<StreamChatGenerics>) => {
  const nonSelfUsers: string[] = [];

  if (!client || !client.user || !typing) {
    return nonSelfUsers;
  }

  const typingKeys = Object.keys(typing);

  typingKeys.forEach((typingKey) => {
    if (!typing[typingKey]) {
      return;
    }

    /** removes own typing events */
    if (client.user?.id === typing[typingKey].user?.id) {
      return;
    }

    const isRegularEvent = !typing[typingKey].parent_id && !thread?.id;
    const isCurrentThreadEvent = typing[typingKey].parent_id === thread?.id;

    /** filters different threads events */
    if (!isRegularEvent && !isCurrentThreadEvent) {
      return;
    }

    const user = typing[typingKey].user?.name || typing[typingKey].user?.id;
    if (user) {
      nonSelfUsers.push(user);
    }
  });

  return nonSelfUsers;
};
