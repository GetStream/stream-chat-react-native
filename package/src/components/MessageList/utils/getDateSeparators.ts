import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type GetDateSeparatorsParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
  userId?: string;
};

export type DateSeparators = {
  [key: string]: Date;
};

export const getDateSeparators = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  params: GetDateSeparatorsParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { hideDateSeparators, messages, userId } = params;
  const dateSeparators: DateSeparators = {};

  if (hideDateSeparators) {
    return dateSeparators;
  }

  const messagesWithoutDeleted = messages.filter((message) => {
    const isMessageTypeDeleted = message.type === 'deleted';
    return !isMessageTypeDeleted || userId === message.user?.id;
  });

  for (let i = 0; i < messagesWithoutDeleted.length; i++) {
    const previousMessage = messagesWithoutDeleted[i - 1];
    const message = messagesWithoutDeleted[i];

    const messageDate = message.created_at.getDay();

    const prevMessageDate = previousMessage ? previousMessage.created_at.getDay() : messageDate;

    if (i === 0 || messageDate !== prevMessageDate) {
      dateSeparators[message.id] = message.created_at;
    }
  }

  return dateSeparators;
};
