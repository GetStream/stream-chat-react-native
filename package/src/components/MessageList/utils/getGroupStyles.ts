import type { DateSeparators } from './getDateSeparators';

import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { GroupType } from '../hooks/useMessageList';

export type GetGroupStylesParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  dateSeparators: DateSeparators;
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
  maxTimeBetweenGroupedMessages?: number;
  noGroupByUser?: boolean;
  userId?: string;
};

export const getGroupStyles = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  params: GetGroupStylesParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    dateSeparators,
    hideDateSeparators,
    maxTimeBetweenGroupedMessages,
    messages,
    noGroupByUser,
    userId,
  } = params;

  if (noGroupByUser) return {};

  const messageGroupStyles: { [key: string]: GroupType[] } = {};

  const messagesFilteredForNonUser = messages.filter((message) => {
    const isMessageTypeDeleted = message.type === 'deleted';
    return !isMessageTypeDeleted || userId === message.user?.id;
  });

  for (let i = 0; i < messagesFilteredForNonUser.length; i++) {
    const previousMessage = messagesFilteredForNonUser[i - 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const message = messagesFilteredForNonUser[i];
    const nextMessage = messagesFilteredForNonUser[i + 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const groupStyles: GroupType[] = [];

    const isPrevMessageTypeDeleted = previousMessage?.type === 'deleted';
    const isNextMessageTypeDeleted = nextMessage?.type === 'deleted';

    const userId = message?.user?.id || null;

    const isTopMessage =
      !previousMessage ||
      previousMessage.type === 'system' ||
      userId !== previousMessage?.user?.id ||
      previousMessage.type === 'error' ||
      !!isPrevMessageTypeDeleted ||
      (!hideDateSeparators && dateSeparators[message.id]) ||
      messageGroupStyles[previousMessage.id]?.includes('bottom');

    const isBottomMessage =
      !nextMessage ||
      nextMessage.type === 'system' ||
      userId !== nextMessage?.user?.id ||
      nextMessage.type === 'error' ||
      !!isNextMessageTypeDeleted ||
      (!hideDateSeparators && dateSeparators[nextMessage.id]) ||
      (maxTimeBetweenGroupedMessages !== undefined &&
        nextMessage.created_at.getTime() - message.created_at.getTime() >
          maxTimeBetweenGroupedMessages);

    /**
     * Add group styles key for top message
     */
    if (isTopMessage) {
      groupStyles.push('top');
    }

    /**
     * Add group styles key for bottom message
     */

    const isMessageTypeDeleted = message.type === 'deleted';
    if (isBottomMessage) {
      /**
       * If the bottom message is also the top, or deleted, or an error,
       * add the key for single message instead of bottom
       */
      if (isTopMessage || isMessageTypeDeleted || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.push('bottom');
      }
    }

    /**
     * Add the key for all non top or bottom messages, if the message is
     * deleted or an error add the key for single otherwise middle
     */
    if (!isTopMessage && !isBottomMessage) {
      if (isMessageTypeDeleted || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('middle');
      }
    }

    if (message.id) {
      messageGroupStyles[message.id] = groupStyles;
    }
  }

  return messageGroupStyles;
};
