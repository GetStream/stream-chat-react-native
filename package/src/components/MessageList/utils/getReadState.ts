import { ChannelState, LocalMessage } from 'stream-chat';

/**
 * Get the number of users who have read the message
 * @param message - The message to get the read state for
 * @param read - The read state of the channel
 * @returns The number of users who have read the message
 */
export const getReadState = (message: LocalMessage, read?: ChannelState['read']) => {
  if (!read) {
    return 0;
  }

  const readState = Object.values(read).reduce((acc, readState) => {
    if (!readState.last_read) {
      return acc;
    }

    if (message.created_at && message.created_at < readState.last_read) {
      return acc + 1;
    }

    return acc;
  }, 0);

  return readState;
};
