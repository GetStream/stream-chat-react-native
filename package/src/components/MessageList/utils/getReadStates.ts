import type { ChannelContextValue } from '../../../contexts/channelContext/ChannelContext';
import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type { DefaultStreamChatGenerics } from '../../../types/types';

export const getReadStates = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  clientUserId: string | undefined,
  messages:
    | PaginatedMessageListContextValue<StreamChatGenerics>['messages']
    | ThreadContextValue<StreamChatGenerics>['threadMessages'],
  read?: ChannelContextValue<StreamChatGenerics>['read'],
) => {
  const readData = messages.reduce((acc, cur) => {
    if (cur.id) {
      acc[cur.id] = false;
    }
    return acc;
  }, {} as { [key: string]: boolean | number });

  const filteredMessagesReversed = messages.filter((msg) => msg.updated_at).reverse();

  if (read) {
    /**
     * Channel read state is stored by user and we only care about users who aren't the client
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { [clientUserId ?? '']: _ignore, ...filteredRead } = read;
    const members = Object.values(filteredRead);

    /**
     * Track number of members who have read previous messages
     */
    let memberReadCount = 0;

    /**
     * Array is in reverse order so newest message is at 0,
     * we find the index of the first message that is older
     * than the last read and then set last read to that, or
     * if there are no newer messages, the first message is
     * last read message.
     */
    for (const message of filteredMessagesReversed) {
      /**
       * If all members are removed then they have read these
       * messages. We do not increment memberReadCount for 1:1
       * chats, so this should be true, not a number in that case.
       */
      if (!members.length) {
        readData[message.id] = memberReadCount || true;
      } else {
        for (const member of members) {
          /**
           * If no last read continue, we can't remove the user
           * because this would mark all messages in a new channel
           * true until at least one other user reads a message.
           */
          if (!member.last_read) {
            continue;
          }

          /**
           * If there there is a last read message add the user
           * to the array of last reads for that message and remove
           * the user from the list of users being checked
           */
          if (message.created_at < member.last_read) {
            /**
             * if this is a direct message the length will be 1
             * as we already deleted the current user from the object
             */
            const numberOfReads = Object.keys(read).length;
            if (numberOfReads === 1) {
              readData[message.id] = true;
            } else {
              const currentMessageReadData = readData[message.id];
              readData[message.id] =
                typeof currentMessageReadData === 'boolean'
                  ? memberReadCount + 1
                  : currentMessageReadData + 1;
            }
            const userIndex = members.findIndex(({ user }) => user.id === member.user?.id);
            if (userIndex !== -1) {
              members.splice(userIndex, 1);
              if (numberOfReads > 1) {
                memberReadCount += 1;
              }
            }
          }
        }

        /**
         * If this is not the last message for a user this will still be
         * set to false. But if other users have read further the number
         * should be how many have read beyond this message.
         */
        if (readData[message.id] === false) {
          readData[message.id] = memberReadCount || false;
        }
      }
    }
  }

  return readData;
};
