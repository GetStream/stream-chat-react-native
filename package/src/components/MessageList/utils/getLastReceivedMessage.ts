import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import { MessageStatusTypes } from '../../../utils/utils';

import type { MessageType } from '../hooks/useMessageList';

export const getLastReceivedMessage = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  messages: MessageType<At, Ch, Co, Ev, Me, Re, Us>[],
) => {
  /**
   * There are no status on dates so they will be skipped
   */
  for (const message of messages) {
    if (
      message?.status === MessageStatusTypes.RECEIVED ||
      message?.status === MessageStatusTypes.SENDING
    ) {
      return message;
    }
  }

  return;
};
