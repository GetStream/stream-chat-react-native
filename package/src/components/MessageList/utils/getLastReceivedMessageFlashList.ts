import { LocalMessage } from 'stream-chat';

import { MessageStatusTypes } from '../../../utils/utils';

export const getLastReceivedMessageFlashList = (messages: LocalMessage[]) => {
  /**
   * There are no status on dates so they will be skipped
   */
  for (let i = messages.length - 1; i >= 0; i--) {
    const message = messages[i];
    if (
      message?.status === MessageStatusTypes.RECEIVED ||
      message?.status === MessageStatusTypes.SENDING
    ) {
      return message;
    }
  }

  return;
};
