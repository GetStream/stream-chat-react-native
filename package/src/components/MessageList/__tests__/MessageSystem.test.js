import React from 'react';

import { cleanup, render, waitFor } from '@testing-library/react-native';

import { ThemeProvider } from '../../../contexts/themeContext/ThemeContext';
import { defaultTheme } from '../../../contexts/themeContext/utils/theme';
import { TranslationProvider } from '../../../contexts/translationContext/TranslationContext';

import { generateMessage, generateStaticMessage } from '../../../mock-builders/generator/message';
import { generateStaticUser } from '../../../mock-builders/generator/user';
import { Streami18n } from '../../../utils/Streami18n';
import { MessageSystem } from '../MessageSystem';

afterEach(cleanup);

describe('MessageSystem', () => {
  it('should render the message system', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const message = generateMessage();
    const { queryByTestId } = render(
      <ThemeProvider style={defaultTheme}>
        <TranslationProvider value={translators}>
          <MessageSystem message={message} />
        </TranslationProvider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(queryByTestId('message-system')).toBeTruthy();
    });
  });

  it('should match the snapshot for message system', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const user = generateStaticUser(0);
    const message = generateStaticMessage('Hello World', { user });
    const { toJSON } = render(
      <ThemeProvider style={defaultTheme}>
        <TranslationProvider value={translators}>
          <MessageSystem message={message} />
        </TranslationProvider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
