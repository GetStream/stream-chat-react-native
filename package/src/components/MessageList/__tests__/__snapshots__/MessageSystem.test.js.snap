// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageSystem should match the snapshot for message system 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      },
      undefined,
      {},
    ]
  }
  testID="message-system"
>
  <View
    style={
      [
        {
          "flex": 1,
          "height": 0.5,
        },
        {
          "backgroundColor": "#ECEBEB",
        },
        {},
      ]
    }
  />
  <View
    style={
      [
        {
          "flex": 3,
          "marginTop": 10,
        },
        {},
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 10,
            "fontWeight": "bold",
            "textAlign": "center",
          },
          {
            "color": "#7A7A7A",
          },
          {},
        ]
      }
    >
      HELLO WORLD
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 10,
            "fontWeight": "bold",
            "textAlign": "center",
          },
          {
            "color": "#7A7A7A",
          },
          {},
        ]
      }
    >
      04/27/2020
    </Text>
  </View>
  <View
    style={
      [
        {
          "flex": 1,
          "height": 0.5,
        },
        {
          "backgroundColor": "#ECEBEB",
        },
        {},
      ]
    }
  />
</View>
`;
